ray-operator/controllers/raycluster_controller.go

```diff
@@ -501,3 +513,168 @@ func (r *RayClusterReconciler) updateStatus(instance *rayiov1alpha1.RayCluster)
 
 	return nil
 }
+
+func (r *RayClusterReconciler) reconcileServiceAccounts(instance *rayiov1alpha1.RayCluster) error {
+	if instance.Spec.EnableInTreeAutoscaling == nil || !*instance.Spec.EnableInTreeAutoscaling {
+		return nil
+	}
+
+	serviceAccounts := corev1.ServiceAccountList{}
+	filterLabels := client.MatchingLabels{common.RayClusterLabelKey: instance.Name}
+	if err := r.List(context.TODO(), &serviceAccounts, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+		return err
+	}
+
+	if serviceAccounts.Items != nil {
+		if len(serviceAccounts.Items) == 1 {
+			r.Log.Info("reconcileServiceAccounts ", "service account found", serviceAccounts.Items[0].Name)
+			return nil
+		}
+
+		// This should never happen.
+		// We add the protection here just in case controller has race issue or user manually create service with same label.
+		if len(serviceAccounts.Items) > 1 {
+			r.Log.Info("reconcileServiceAccounts ", "Duplicates service account found", len(serviceAccounts.Items))
+			return nil
+		}
+	}
+
+	// Create head service if there's no existing one in the cluster.
+	if serviceAccounts.Items == nil || len(serviceAccounts.Items) == 0 {
+		serviceAccount, err := common.BuildServiceAccount(instance)
+		if err != nil {
+			return err
+		}
+
+		// making sure the name is valid
+		serviceAccount.Name = utils.CheckName(serviceAccount.Name)
+		// Set controller reference
+		if err := controllerutil.SetControllerReference(instance, serviceAccount, r.Scheme); err != nil {
+			return err
+		}
+
+		if errSvc := r.Create(context.TODO(), serviceAccount); errSvc != nil {
+			if errors.IsAlreadyExists(errSvc) {
+				log.Info("Pod service account already exist,no need to create")
+				return nil
+			}
+			log.Error(errSvc, "Pod Service Account create error!", "Pod.ServiceAccount.Error", errSvc)
+			return errSvc
+		}
+		log.Info("Pod ServiceAccount created successfully", "service account name", serviceAccount.Name)
+		r.Recorder.Eventf(instance, v1.EventTypeNormal, "Created", "Created service account %s", serviceAccount.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func (r *RayClusterReconciler) reconcileRole(instance *rayiov1alpha1.RayCluster) error {
```

Comment:
For the name of the method, maybe "reconcileAutoscalerRole" or something like that?

Reply:
Agree. At this moment, something like `reconcileAutoscalerRole ` makes more sense. Later when it manages more roles, we may rename to `reconcileRole` and handle different roles inside the method just like we did for head/work pod/services